File: game.py

import math
import pygame
import pymunk
from pymunk import Vec2d
from objects import Wall, Ball, Rectangle
from rocket import Rocket
from tests import ALL_TESTS


class RocketGame():

    # Settings
    FRAME_WIDTH = 1600  # in px
    FRAME_HEIGHT = 900  # in px
    GRAVITY = 600
    DRAG = 1
    DT = 1. / 1000.  # in seconds
    FPS = 50
    SAVE_IMG = False
    # ZOOM = 1  # in px / meter

    def __init__(self):

        # screen and runtime
        self.screen = pygame.display.set_mode(
            (self.FRAME_WIDTH, self.FRAME_HEIGHT), pygame.RESIZABLE)
        self.clock = pygame.time.Clock()
        self.time = 0

        # game state variables
        self.running = True
        self.run_physics = True

        # physics stuff
        self.space = pymunk.Space()
        self.space.gravity = 0.0, -self.GRAVITY

        # game objects
        self.objects = []
        self.rocket = None

        # list of events and pressed keys
        self.events = []
        self.pressed_keys = []

        # add ground
        for n in range(4):
            self.add_object(
                Wall(self.space, Vec2d(-1e9, 2*n), Vec2d(1e9, 2*n)))
        # add the rocket
        self.add_new_rocket()

        # storage for a unit test, if performing a test
        self.test = None

    # add an object to the game and return it
    def add_object(self, obj):
        if obj not in self.objects:
            self.objects.append(obj)
        return obj

    # remove an object from the game
    def remove_object(self, obj):
        if obj in self.objects:
            self.objects.remove(obj)

    # add a new rocket
    def add_new_rocket(self):
        # remove the previous rocket
        if self.rocket is not None:
            self.remove_object(self.rocket)
        # create new rocket
        self.rocket = Rocket(self.space, 0, 100)
        # add it to the game
        self.add_object(self.rocket)

        # # add some component
        # rpos = self.rocket.body.position + Vec2d(20, 0)
        # rect = Rectangle(self.space, *rpos, 10, 10)
        # self.add_object(rect)
        # constraint = pymunk.PinJoint(self.rocket.body, rect.body, (20, 0), (0, 0))
        # self.space.add(constraint)

        # # add some component
        # rpos = self.rocket.body.position + Vec2d(-20, 0)
        # rect = Rectangle(self.space, *rpos, 10, 10)
        # self.add_object(rect)
        # constraint = pymunk.PinJoint(self.rocket.body, rect.body, (-20, 0), (0, 0))
        # self.space.add(constraint)

    # update the physics of the game and each object
    def update_physics(self):
        if self.run_physics:
            # update the external forces for each object
            for obj in self.objects:
                obj.update_drag()
                obj.update_forces()
            # perform time steps
            self.space.step(self.DT)
            # update time
            self.time += self.DT

    # process the controls of the game (events, pressed keys, etc.)
    def handle_controls(self):

        self.events = pygame.event.get()
        self.pressed_keys = pygame.key.get_pressed()

        for event in self.events:
            if event.type == pygame.QUIT:
                self.running = False
            elif event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE:
                self.running = False
            elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
                x, y = self.screen2pos(event.pos)
                self.add_object(Ball(self.space, x, y))
            elif event.type == pygame.KEYDOWN and event.key == pygame.K_r:
                self.add_new_rocket()
            elif event.type == pygame.KEYDOWN and event.key == pygame.K_p:
                self.run_physics = not self.run_physics
            elif event.type == pygame.KEYDOWN and event.key == pygame.K_t:
                if self.test is None:
                    self.run_tests()
            elif event.type == pygame.VIDEORESIZE:
                self.FRAME_WIDTH = event.w
                self.FRAME_HEIGHT = event.h
                self.screen = pygame.display.set_mode(
                    (self.FRAME_WIDTH, self.FRAME_HEIGHT), pygame.RESIZABLE)

        # handle unit tests
        if self.test is not None:
            if self.test.is_finished():
                self.test = self.test.next_test
                if self.test is not None:
                    self.test.start(self)

        # update rocket controls
        self.rocket.handle_controls(self)

    # draw the game
    def draw(self):
        # Clear screen
        self.screen.fill(pygame.Color("white"))
        # Draw background
        x, y = self.rocket.body.position
        color = (240, 240, 240)
        dist = 700
        offsx, offsy = -x % dist, y % dist
        for i in range(-1, 5):
            pygame.draw.rect(self.screen, color,
                             (offsx + i*dist, 0, dist/20, self.FRAME_HEIGHT))
            pygame.draw.rect(self.screen, color,
                             (0, offsy + i*dist, self.FRAME_WIDTH, dist/20))
        # Draw objects
        for obj in self.objects:
            obj.draw(self)
        # Display some text
        font = pygame.font.Font(None, 24)

        thrust = self.rocket.engine.thrust/self.rocket.engine.MAX_THRUST
        twr = self.rocket.twr()
        h = self.length_unit(self.rocket.body.position.y)
        v = self.velocity_unit(self.rocket.body.velocity.y)
        sas = self.rocket.pilot.sas_mode
        text = f"""Thrust: {thrust:3.0%}
TWR: {twr:.2f}
Height: {h}
Velocity: {v}
SAS: {sas}

Controls:
---------
Up/Down - throttle
Left/Right - thrust vector control
space - start/stop engine
0-4 - switch SAS mode
mouse - add obstacle
P - pause
R - restart
T - run tests
"""
        y = 5
        for line in text.splitlines():
            text = font.render(line, True, pygame.Color("black"))
            self.screen.blit(text, (5, y))
            y += 20

    # transform global coordinates to screen coordinates
    def pos2screen(self, pos):
        x = pos.x + self.FRAME_WIDTH / 2.
        y = pos.y - self.FRAME_HEIGHT / 2.
        if self.rocket is not None:
            x -= self.rocket.body.position.x
            y -= self.rocket.body.position.y
        return int(x), -int(y)

    # transform screen coordinates to global coordinates
    def screen2pos(self, pos):
        x = pos[0] - self.FRAME_WIDTH / 2.
        y = -pos[1] + self.FRAME_HEIGHT / 2.
        if self.rocket is not None:
            x += self.rocket.body.position.x
            y += self.rocket.body.position.y
        return Vec2d(x, y)

    # convert lengthlength_unitto string with units
    def length_unit(self, h, precision=1):
        if abs(h) < 1e3:
            return f"{h:,.{precision}f}m"
        ly = 9.461e15
        if abs(h) < 0.1*ly:
            return f"{h/1e3:,.{precision}f}km"
        else:
            return f"{h/ly:,.{precision}f}ly"

    # convert velocity to string with units
    def velocity_unit(self, v, precision=1):
        if abs(v) < 1e3:
            return f"{v:,.{precision}f}m/s"
        if abs(v) > 1e3:
            return f"{v/1e3:,.{precision}f}km/s"

    # Performs landing tests for different scenarios and measures how well the rocket performs.
    def run_tests(self):
        for n in range(len(ALL_TESTS)-1):
            ALL_TESTS[n].next_test = ALL_TESTS[n+1]
        ALL_TESTS[0].start(self)

File: src/game.ts
import Matter from "matter-js";
import { GameObject, Ball, Wall, Game as IGame } from "./objects";
import { Rocket } from "./rocket";

/**
 * Main game class for the rocket simulation.
 */
export class RocketGame implements IGame {
  readonly FRAME_WIDTH = 1600;
  readonly FRAME_HEIGHT = 900;
  readonly GRAVITY = 600;
  readonly FPS = 50;
  readonly DT = 1 / this.FPS;

  canvas: HTMLCanvasElement;
  ctx: CanvasRenderingContext2D;
  engine: Matter.Engine;
  world: Matter.World;

  objects: GameObject[] = [];
  rocket: Rocket | null = null;

  running = true;
  run_physics = true;
  time = 0;

  pressed_keys: { [key: string]: boolean } = {};

  events: { type: string; code: string }[] = [];

  test: {
    is_finished: () => boolean;

    next_test: { start: (game: RocketGame) => void } | null;

    start: (game: RocketGame) => void;
  } | null = null;

  constructor(canvas: HTMLCanvasElement) {
    this.canvas = canvas;
    const ctx = canvas.getContext("2d");
    if (!ctx) {
      throw new Error("Could not get canvas context");
    }
    this.ctx = ctx;

    this.canvas.width = this.FRAME_WIDTH;
    this.canvas.height = this.FRAME_HEIGHT;

    this.engine = Matter.Engine.create();
    this.world = this.engine.world;
    this.engine.gravity.y = -1;
    this.engine.gravity.scale = (1 * this.GRAVITY) / 9.81;

    this.add_ground();
    this.add_new_rocket();

    window.addEventListener("keydown", (e) => {
      this.pressed_keys[e.code] = true;
      this.events.push({ type: "keydown", code: e.code });
    });
    window.addEventListener("keyup", (e) => {
      this.pressed_keys[e.code] = false;
    });
    window.addEventListener("mousedown", (e) => {
      const rect = this.canvas.getBoundingClientRect();
      const x = e.clientX - rect.left;
      const y = e.clientY - rect.top;
      const pos = this.screen2pos({ x, y });
      this.add_object(new Ball(this.world, pos.x, pos.y));
    });

    // Handle window resize
    window.addEventListener("resize", () => this.handle_resize());
  }

  handle_resize(): void {
    // Simple resize logic
    // For now keep it fixed or adjust canvas size
  }

  add_ground(): void {
    for (let n = 0; n < 4; n++) {
      this.add_object(new Wall(this.world, { x: -1e6, y: 2 * n }, { x: 1e6, y: 2 * n }));
    }
  }

  add_new_rocket(): void {
    if (this.rocket) {
      this.remove_object(this.rocket);
    }
    this.rocket = new Rocket(this.world, 0, 200);
    this.rocket.space = { gravity: { x: 0, y: -this.GRAVITY } };
    this.add_object(this.rocket);
  }

  add_object(obj: GameObject): GameObject {
    this.objects.push(obj);
    return obj;
  }

  remove_object(obj: GameObject): void {
    const index = this.objects.indexOf(obj);
    if (index !== -1) {
      this.objects.splice(index, 1);
      if (obj.body) {
        Matter.World.remove(this.world, obj.body);
      }
    }
  }

  update_physics(): void {
    if (this.run_physics) {
      for (const obj of this.objects) {
        // obj.update_drag();
        obj.update_forces();
      }
      Matter.Engine.update(this.engine, this.DT);
      this.time += this.DT;
    }
  }

  handle_controls(): void {
    // Process internal events
    if (this.rocket) {
      this.rocket.handle_controls(this);
    }

    for (const event of this.events) {
      if (event.type === "keydown") {
        if (event.code === "KeyR") {
          this.add_new_rocket();
        }
        if (event.code === "KeyP") {
          this.run_physics = !this.run_physics;
        }
        if (event.code === "KeyT") {
          this.run_tests();
        }
        if (event.code === "KeyA" && this.rocket) {
          this.rocket.airbrakes_enabled = !this.rocket.airbrakes_enabled;
        }
      }
    }

    // handle unit tests
    if (this.test !== null) {
      if (this.test.is_finished()) {
        this.test = this.test.next_test;
        if (this.test !== null) {
          this.test.start(this);
        }
      }
    }

    this.events = [];
  }

  run_tests(): void {
    import("./tests").then(({ ALL_TESTS }) => {
      for (let n = 0; n < ALL_TESTS.length - 1; n++) {
        ALL_TESTS[n].next_test = ALL_TESTS[n + 1];
      }
      this.test = ALL_TESTS[0];
      this.test.start(this);
    });
  }

  draw(): void {
    // Clear
    this.ctx.fillStyle = "white";
    this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);

    // Grid/Background
    if (this.rocket && this.rocket.body) {
      const { x, y } = this.rocket.body.position;
      const color = "rgb(240, 240, 240)";
      const dist = 700;
      const offsx = -x % dist;
      const offsy = y % dist;
      this.ctx.strokeStyle = color;
      this.ctx.lineWidth = dist / 20;

      for (let i = -1; i < 5; i++) {
        // Vertical lines
        this.ctx.beginPath();
        this.ctx.moveTo(offsx + i * dist, 0);
        this.ctx.lineTo(offsx + i * dist, this.FRAME_HEIGHT);
        this.ctx.stroke();

        // Horizontal lines
        this.ctx.beginPath();
        this.ctx.moveTo(0, offsy + i * dist);
        this.ctx.lineTo(this.FRAME_WIDTH, offsy + i * dist);
        this.ctx.stroke();
      }
    }

    // Objects
    for (const obj of this.objects) {
      obj.draw(this);
    }

    // Text
    if (this.rocket) {
      this.ctx.fillStyle = "black";
      this.ctx.font = "18px monospace";
      const thrust = this.rocket.engine.thrust / this.rocket.engine.MAX_THRUST;
      const twr = this.rocket.twr();
      const h = this.length_unit(this.rocket.body!.position.y);
      const v = this.velocity_unit(this.rocket.body!.velocity.y);
      const sas = this.rocket.pilot.sas_mode;

      const lines = [
        `Thrust: ${(thrust * 100).toFixed(0)}%`,
        `TWR: ${twr.toFixed(2)}`,
        `Height: ${h}`,
        `Velocity: ${v}`,
        `SAS: ${sas}`,
        "",
        "Controls:",
        "---------",
        "Up/Down - throttle",
        "Left/Right - thrust vector control",
        "space - start/stop engine",
        "0-4 - switch SAS mode",
        "mouse - add obstacle",
        "A - airbrakes",
        "P - pause",
        "R - restart",
        "T - run tests",
      ];

      let y = 25;
      for (const line of lines) {
        this.ctx.fillText(line, 10, y);
        y += 20;
      }
    }
  }

  pos2screen(pos: Matter.Vector): { x: number; y: number } {
    let x = pos.x + this.FRAME_WIDTH / 2;
    let y = pos.y - this.FRAME_HEIGHT / 2;
    if (this.rocket && this.rocket.body) {
      x -= this.rocket.body.position.x;
      y -= this.rocket.body.position.y;
    }
    return { x, y: -y };
  }

  screen2pos(pos: { x: number; y: number }): Matter.Vector {
    let x = pos.x - this.FRAME_WIDTH / 2;
    let y = -pos.y + this.FRAME_HEIGHT / 2;
    if (this.rocket && this.rocket.body) {
      x += this.rocket.body.position.x;
      y += this.rocket.body.position.y;
    }
    return { x, y };
  }

  length_unit(h: number): string {
    if (Math.abs(h) < 1e3) {
      return `${h.toFixed(1)}m`;
    }
    const ly = 9.461e15;
    if (Math.abs(h) < 0.1 * ly) {
      return `${(h / 1e3).toFixed(1)}km`;
    }
    return `${(h / ly).toFixed(1)}ly`;
  }

  velocity_unit(v: number): string {
    if (Math.abs(v) < 1e3) {
      return `${v.toFixed(1)}m/s`;
    }
    return `${(v / 1e3).toFixed(1)}km/s`;
  }
}
